File: daytona-terminal/backend/app/api/ai.py
"""AI Assistant API routes."""

from fastapi import APIRouter, HTTPException, status

from app.models.ai_assistant import (
    AIAssistant,
    AIAssistantType,
    AICommandRequest,
    AICommandResponse,
    AIConversation,
)
from app.services.ai_assistant import AIAssistantError, get_ai_assistant_service
from app.services.terminal import get_terminal_service

router = APIRouter(prefix="/ai", tags=["ai"])


@router.get("/assistants", response_model=list[AIAssistant])
async def list_assistants():
    """List available AI assistants."""
    service = get_ai_assistant_service()
    return service.list_assistants()


@router.get("/assistants/{assistant_type}", response_model=AIAssistant)
async def get_assistant(assistant_type: str):
    """Get an AI assistant by type."""
    service = get_ai_assistant_service()
    assistant = service.get_assistant(assistant_type)
    if not assistant:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Assistant not found: {assistant_type}",
        )
    return assistant


@router.post("/conversations", response_model=AIConversation)
async def create_conversation(
    session_id: str,
    assistant_type: AIAssistantType = AIAssistantType.CLAUDE,
    system_prompt: str | None = None,
):
    """Create a new AI conversation for a terminal session."""
    service = get_ai_assistant_service()
    
    # Verify session exists
    terminal_service = get_terminal_service()
    session = await terminal_service.get_session(session_id)
    if not session:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Session not found: {session_id}",
        )
    
    conversation = await service.create_conversation(
        session_id, assistant_type, system_prompt
    )
    return conversation


@router.get("/conversations/{conversation_id}", response_model=AIConversation)
async def get_conversation(conversation_id: str):
    """Get a conversation by ID."""
    service = get_ai_assistant_service()
    conversation = await service.get_conversation(conversation_id)
    if not conversation:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Conversation not found: {conversation_id}",
        )
    return conversation


@router.post("/conversations/{conversation_id}/message", response_model=AICommandResponse)
async def send_message(conversation_id: str, request: AICommandRequest):
    """Send a message to the AI assistant."""
    service = get_ai_assistant_service()
    terminal_service = get_terminal_service()
    
    # Get conversation
    conversation = await service.get_conversation(conversation_id)
    if not conversation:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Conversation not found: {conversation_id}",
        )
    
    # Get terminal context if requested
    terminal_context = None
    if request.include_terminal_context:
        terminal_context = terminal_service.get_terminal_context(
            conversation.session_id
        )
    
    try:
        response = await service.send_message(
            conversation_id, request, terminal_context
        )
        return response
    except AIAssistantError as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=str(e),
        )


@router.delete("/conversations/{conversation_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_conversation(conversation_id: str):
    """Delete a conversation."""
    service = get_ai_assistant_service()
    success = await service.delete_conversation(conversation_id)
    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Conversation not found: {conversation_id}",
        )


@router.post("/conversations/{conversation_id}/clear", status_code=status.HTTP_204_NO_CONTENT)
async def clear_conversation(conversation_id: str):
    """Clear messages from a conversation."""
    service = get_ai_assistant_service()
    success = await service.clear_conversation(conversation_id)
    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Conversation not found: {conversation_id}",
        )

File: daytona-terminal/backend/src/api/ai.ts
import type { FastifyPluginAsync } from 'fastify';
import { z } from 'zod';
import { aiService } from '../services/ai.service.js';
import { terminalService } from '../services/terminal.service.js';
import { AICommandRequestSchema, AIAssistantType } from '../models/ai.js';

export const aiRoutes: FastifyPluginAsync = async (fastify) => {
  // List available AI assistants
  fastify.get('/assistants', async () => {
    return aiService.listAssistants();
  });

  // Get specific assistant
  fastify.get('/assistants/:type', async (request, reply) => {
    const { type } = request.params as { type: string };
    const assistant = aiService.getAssistant(type);
    if (!assistant) {
      reply.code(404);
      return { error: `Assistant not found: ${type}` };
    }
    return assistant;
  });

  // Create conversation
  fastify.post('/conversations', async (request, reply) => {
    const { sessionId, assistantType = 'claude', systemPrompt } = request.body as {
      sessionId: string;
      assistantType?: string;
      systemPrompt?: string;
    };

    if (!sessionId) {
      reply.code(400);
      return { error: 'sessionId is required' };
    }

    // Verify session exists
    const session = await terminalService.getSession(sessionId);
    if (!session) {
      reply.code(404);
      return { error: `Session not found: ${sessionId}` };
    }

    try {
      const validType = AIAssistantType.parse(assistantType);
      const conversation = await aiService.createConversation(sessionId, validType, systemPrompt);
      reply.code(201);
      return conversation;
    } catch (error) {
      if (error instanceof z.ZodError) {
        reply.code(400);
        return { error: 'Invalid assistant type' };
      }
      throw error;
    }
  });

  // Get conversation
  fastify.get('/conversations/:id', async (request, reply) => {
    const { id } = request.params as { id: string };
    const conversation = await aiService.getConversation(id);
    if (!conversation) {
      reply.code(404);
      return { error: `Conversation not found: ${id}` };
    }
    return conversation;
  });

  // Send message
  fastify.post('/conversations/:id/message', async (request, reply) => {
    const { id } = request.params as { id: string };

    const conversation = await aiService.getConversation(id);
    if (!conversation) {
      reply.code(404);
      return { error: `Conversation not found: ${id}` };
    }

    try {
      const body = AICommandRequestSchema.parse(request.body);

      // Get terminal context if requested
      let terminalContext: string | undefined;
      if (body.includeTerminalContext) {
        terminalContext = terminalService.getContext(conversation.sessionId, 10000);
      }

      const response = await aiService.sendMessage(id, body, terminalContext);
      return response;
    } catch (error) {
      if (error instanceof z.ZodError) {
        reply.code(400);
        return { error: 'Validation error', details: error.errors };
      }
      if (error instanceof Error) {
        reply.code(500);
        return { error: error.message };
      }
      throw error;
    }
  });

  // Delete conversation
  fastify.delete('/conversations/:id', async (request, reply) => {
    const { id } = request.params as { id: string };
    const success = await aiService.deleteConversation(id);
    if (!success) {
      reply.code(404);
      return { error: `Conversation not found: ${id}` };
    }
    reply.code(204);
    return;
  });

  // Clear conversation
  fastify.post('/conversations/:id/clear', async (request, reply) => {
    const { id } = request.params as { id: string };
    const success = await aiService.clearConversation(id);
    if (!success) {
      reply.code(404);
      return { error: `Conversation not found: ${id}` };
    }
    reply.code(204);
    return;
  });

  // Quick analyze (for inline error suggestions)
  fastify.post('/analyze', async (request, reply) => {
    const { sessionId, output } = request.body as {
      sessionId: string;
      output: string;
    };

    if (!sessionId || !output) {
      reply.code(400);
      return { error: 'sessionId and output are required' };
    }

    const result = await aiService.quickAnalyze(sessionId, output);
    return result || { suggestion: null, commands: [] };
  });

  // Get completions
  fastify.post('/completions', async (request, reply) => {
    const { sessionId, partial } = request.body as {
      sessionId: string;
      partial: string;
    };

    if (!sessionId) {
      reply.code(400);
      return { error: 'sessionId is required' };
    }

    const completions = await aiService.getCompletions(sessionId, partial || '');
    return { completions };
  });
};
